resolver/resolver.go

```diff
@@ -354,111 +347,117 @@ func (res *Resolver) HandleNonMesos(w dns.ResponseWriter, r *dns.Msg) {
 // question with resource answer(s)
 // it can handle {A, SRV, ANY}
 func (res *Resolver) HandleMesos(w dns.ResponseWriter, r *dns.Msg) {
-	var err error
+	defer logging.CurLog.MesosRequests.Inc()
 
-	dom := strings.ToLower(cleanWild(r.Question[0].Name))
-	qType := r.Question[0].Qtype
-
-	m := new(dns.Msg)
-	m.Authoritative = true
-	m.RecursionAvailable = res.config.RecurseOn
+	m := &dns.Msg{MsgHdr: dns.MsgHdr{
+		Authoritative:      true,
+		RecursionAvailable: res.config.RecurseOn,
+	}}
 	m.SetReply(r)
 
 	rs := res.records()
-
-	// SRV requests
-	if (qType == dns.TypeSRV) || (qType == dns.TypeANY) {
-		for _, srv := range rs.SRVs[dom] {
-			rr, err := res.formatSRV(r.Question[0].Name, srv)
-			if err != nil {
-				logging.Error.Println(err)
-			} else {
-				m.Answer = append(m.Answer, rr)
-				// return one corresponding A record add additional info
-				host := strings.Split(srv, ":")[0]
-				if len(rs.As[host]) != 0 {
-					rr, err := res.formatA(host, rs.As[host][0])
-					if err != nil {
-						logging.Error.Println(err)
-					} else {
-						m.Extra = append(m.Extra, rr)
-					}
-				}
-
-			}
-		}
+	name := strings.ToLower(cleanWild(r.Question[0].Name))
+	switch r.Question[0].Qtype {
+	case dns.TypeSRV, dns.TypeANY:
+		res.handleSRV(rs, name, m, r)
+	case dns.TypeA:
+		res.handleA(rs, name, m)
+	case dns.TypeSOA:
+		res.handleSOA(m, r)
+	case dns.TypeNS:
+		res.handleNS(m, r)
+	}
+
+	if len(m.Answer) == 0 {
+		res.handleEmpty(rs, name, m, r)
+	} else {
+		shuffleAnswers(m.Answer)
+		logging.CurLog.MesosSuccess.Inc()
 	}
 
-	// A requests
-	if (qType == dns.TypeA) || (qType == dns.TypeANY) {
-		for _, a := range rs.As[dom] {
-			rr, err := res.formatA(dom, a)
-			if err != nil {
-				logging.Error.Println(err)
-			} else {
-				m.Answer = append(m.Answer, rr)
-			}
+	reply(w, m)
+}
 
+func (res *Resolver) handleSRV(rs *records.RecordGenerator, name string, m, r *dns.Msg) {
+	for _, srv := range rs.SRVs[name] {
+		srvRR, err := res.formatSRV(r.Question[0].Name, srv)
+		if err != nil {
+			logging.CurLog.MesosFailed.Inc()
+			logging.Error.Println(err)
+			continue
 		}
-	}
-
-	// SOA requests
-	if (qType == dns.TypeSOA) || (qType == dns.TypeANY) {
-		rr, err := res.formatSOA(r.Question[0].Name)
+		m.Answer = append(m.Answer, srvRR)
+		// return one corresponding A record add additional info
+		host := strings.Split(srv, ":")[0]
+		if len(rs.As[host]) == 0 {
+			continue
+		}
+		aRR, err := res.formatA(host, rs.As[host][0])
 		if err != nil {
+			logging.CurLog.MesosFailed.Inc()
```

Comment:
This changes the metric. In the old code the counter only counted the requests. Now it's iterating over the srv's.